quixstreams/app.py

```diff
@@ -75,6 +75,78 @@ def __call__(
     ) -> TopicManager: ...
 
 
+class RunTracker:
+    __slots__ = ("running", "_start_time", "_timeout", "_stop_checker", "_run_count")
+
+    _timeout_default = 0.0
+
+    def __init__(self):
+        """
+        Tracks the runtime status of an Application, along with managing variables
+        associated with stopping the app based on a timeout
+        """
+        self.running: bool = False
+        self._start_time: float = time.monotonic()
+        self._timeout: float = self._timeout_default
+        self._run_count: int = 0
+        self._stop_checker: Optional[Callable[[], bool]] = None
+
+    @property
+    def run_count(self) -> int:
+        """
+        Tracks number of times app.run() has been called.
+        Mostly only here for tracking the initial run.
+        :return:
+        """
+        return self._run_count
+
+    def update_status(self):
+        """
+        Trigger stop if its condition is met.
+        """
+        # most optimal way to keep performance with typical operation (no timeout)
+        if stopper := self._stop_checker:
+            if stopper():
+                self.stop()
+
+    def start(self):
+        """
+        Called as part of app.run() to initialize/reset stop conditions.
+        """
+        self.running = True
+        self._run_count += 1
+        self.reset_tracking()
+        if self._timeout:
+            self._stop_checker = self._at_timeout
```

Comment:
I suggest to maintain an additional boolean to indicate the activity of a stop checker. Elaborated in a different comment.

```suggestion
            self._stop_checker = self._at_timeout
            self._stop_checker_active = True
```